Sources/BSON/BSONDocument.swift

```diff
@@ -339,7 +303,53 @@ extension BSONDocument: BSONValue {
     }
 
     internal func write(to buffer: inout ByteBuffer) {
-        var doc = ByteBuffer(self._buffer.readableBytesView)
+        var doc = ByteBuffer(self.storage.buffer.readableBytesView)
         buffer.writeBuffer(&doc)
     }
 }
+
+internal struct BSONDocumentStorage {
+    internal var buffer: ByteBuffer
+    internal init(_ buffer: ByteBuffer) { self.buffer = buffer }
+
+    /// Appends element to underlying BSON bytes, returns the size of the element appended: type + key + value
+    @discardableResult internal static func append(key: String, value: BSON, to buffer: inout ByteBuffer) -> Int {
```

Comment:
this can drop the `to` argument and no longer be static, since now the storage owns the buffer.